File: lib/scanning_screen.dart
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:ningapi/app_colors.dart';
import 'package:flutter_animate/flutter_animate.dart';
import 'package:ningapi/services/currency_recognition_service.dart';
import 'package:ningapi/services/face_detection_service.dart';
import 'package:ningapi/services/tts_service.dart';
import 'package:ningapi/results_screen.dart';

class ScanningScreen extends StatefulWidget {
  final File imageFile;

  const ScanningScreen({super.key, required this.imageFile});

  @override
  State<ScanningScreen> createState() => _ScanningScreenState();
}

class _ScanningScreenState extends State<ScanningScreen> with SingleTickerProviderStateMixin {
  late AnimationController _animationController;
  late Animation<double> _animation;
  final TtsService _tts = TtsService();
  bool _isProcessing = false;
  bool _isDisposed = false;

  @override
  void initState() {
    super.initState();
    _animationController = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 1500),
    )..repeat();

    _animation = Tween<double>(begin: 0, end: 1).animate(
      CurvedAnimation(parent: _animationController, curve: Curves.easeInOut),
    );

    // Utiliser speakAsync pour ne pas bloquer
    _tts.speakAsync("Analyse en cours");
    _performRecognition();
  }

  Future<void> _performRecognition() async {
    if (_isProcessing || _isDisposed) return;
    _isProcessing = true;

    try {
      // 1. D'abord, verifier s'il s'agit d'un visage
      final hasFace = await FaceDetectionService().hasFace(widget.imageFile);
      
      if (_isDisposed || !mounted) return;

      if (hasFace) {
        // Si c'est un visage, on avertit l'utilisateur
        await Future.delayed(const Duration(milliseconds: 500));
        await _tts.speak("Attention. Ceci n'est pas un billet, c'est un visage. Veuillez scanner un billet.");
        
        if (mounted && !_isDisposed) {
          // Retour a l'ecran precedent apres l'annonce
          await Future.delayed(const Duration(seconds: 2));
          if (mounted && !_isDisposed) Navigator.pop(context);
        }
        return;
      }

      // 2. Si pas de visage, on lance la reconnaissance de billet
      final result = await CurrencyRecognitionService().recognizeCurrency(widget.imageFile);

      if (_isDisposed || !mounted) return;

      // Petit delai pour l'UX
      await Future.delayed(const Duration(milliseconds: 500));

      if (mounted && !_isDisposed) {
        // Arreter le TTS avant navigation
        await _tts.stop();

        // Navigation avec Navigator - remplacer l'ecran actuel
        Navigator.of(context).pushReplacement(
          MaterialPageRoute(
            builder: (_) => ResultsScreen(
              imageFile: widget.imageFile,
              currencyResult: result,
            ),
          ),
        );
      }
    } catch (e) {
      debugPrint('Erreur reconnaissance: $e');
      if (mounted && !_isDisposed) {
        await _tts.speak("Erreur lors de l'analyse. Veuillez reessayer.");
        await Future.delayed(const Duration(seconds: 2));
        if (mounted && !_isDisposed) {
          Navigator.pop(context);
        }
      }
    } finally {
      _isProcessing = false;
    }
  }

  @override
  void dispose() {
    _isDisposed = true;
    _animationController.dispose();
    _tts.stop();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: AppColors.backgroundLight,
      body: SafeArea(
        child: Padding(
          padding: const EdgeInsets.all(24.0),
          child: Column(
            children: [
              const Spacer(),

              // Image with scan effect
              Container(
                constraints: const BoxConstraints(maxWidth: 320, maxHeight: 420),
                decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(24),
                  border: Border.all(color: AppColors.accent, width: 3),
                  boxShadow: [
                    BoxShadow(
                      color: AppColors.accent.withOpacity(0.25),
                      blurRadius: 40,
                      spreadRadius: 5,
                    ),
                  ],
                ),
                clipBehavior: Clip.hardEdge,
                child: Stack(
                  children: [
                    SizedBox.expand(
                      child: Image.file(widget.imageFile, fit: BoxFit.cover),
                    ),

                    // Light overlay
                    Container(color: Colors.white.withOpacity(0.1)),

                    // Turquoise scan line
                    AnimatedBuilder(
                      animation: _animation,
                      builder: (context, child) {
                        return Positioned(
                          top: 420 * _animation.value,
                          left: 0,
                          right: 0,
                          child: Container(
                            height: 4,
                            decoration: BoxDecoration(
                              gradient: LinearGradient(
                                colors: [
                                  Colors.transparent,
                                  AppColors.accent,
                                  AppColors.accentLight,
                                  AppColors.accent,
                                  Colors.transparent,
                                ],
                              ),
                              boxShadow: [
                                BoxShadow(
                                  color: AppColors.accent,
                                  blurRadius: 20,
                                  spreadRadius: 4,
                                ),
                              ],
                            ),
                          ),
                        );
                      },
                    ),
                  ],
                ),
              )
                  .animate()
                  .fadeIn(duration: 400.ms),

              const SizedBox(height: 50),

              // Status card
              Container(
                padding: const EdgeInsets.all(28),
                decoration: BoxDecoration(
                  color: AppColors.backgroundCard,
                  borderRadius: BorderRadius.circular(20),
                  border: Border.all(color: AppColors.borderColor),
                  boxShadow: [
                    BoxShadow(
                      color: Colors.black.withOpacity(0.05),
                      blurRadius: 20,
                      offset: const Offset(0, 5),
                    ),
                  ],
                ),
                child: Column(
                  children: [
                    // Animated icon
                    Container(
                      width: 55,
                      height: 55,
                      decoration: BoxDecoration(
                        color: AppColors.accent,
                        shape: BoxShape.circle,
                      ),
                      child: const Icon(Icons.search, color: AppColors.textLight, size: 28),
                    )
                        .animate(onPlay: (c) => c.repeat())
                        .scale(begin: const Offset(1, 1), end: const Offset(1.1, 1.1), duration: 600.ms)
                        .then()
                        .scale(begin: const Offset(1.1, 1.1), end: const Offset(1, 1), duration: 600.ms),

                    const SizedBox(height: 20),

                    const Text(
                      "Analyse...",
                      style: TextStyle(
                        color: AppColors.textDark,
                        fontSize: 32,
                        fontWeight: FontWeight.bold,
                      ),
                    ),
                    const SizedBox(height: 16),
                    const Text(
                      "Veuillez patienter",
                      style: TextStyle(color: AppColors.textGray, fontSize: 18),
                    ),
                    const SizedBox(height: 32),

                    // Progress bar
                    ClipRRect(
                      borderRadius: BorderRadius.circular(4),
                      child: const LinearProgressIndicator(
                        minHeight: 5,
                        backgroundColor: AppColors.borderColor,
                        valueColor: AlwaysStoppedAnimation<Color>(AppColors.accent),
                        value: null,
                      ),
                    ),
                  ],
                ),
              ),

              const Spacer(),

              // Cancel
              TextButton(
                onPressed: () => Navigator.of(context).pop(),
                child: const Text(
                  "Annuler",
                  style: TextStyle(color: AppColors.textGray, fontSize: 16),
                ),
              ),
              const SizedBox(height: 20),
            ],
          ),
        ),
      ),
    );
  }
}

File: .OTLIN/ui/screens/ScanningScreen.kt
package com.ningapi.ui.screens

import android.graphics.BitmapFactory
import androidx.compose.animation.core.*
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavController
import com.ningapi.services.CurrencyRecognitionService
import com.ningapi.services.FaceDetectionService
import com.ningapi.services.TtsService
import com.ningapi.ui.theme.AppColors
import kotlinx.coroutines.delay
import java.io.File

@Composable
fun ScanningScreen(
    navController: NavController,
    imageFile: File
) {
    val context = LocalContext.current
    val ttsService = remember { TtsService.getInstance(context) }
    
    var isProcessing by remember { mutableStateOf(false) } 
    
    // Animation
    val infiniteTransition = rememberInfiniteTransition(label = "scan")
    val scanLinePosition by infiniteTransition.animateFloat(
        initialValue = 0f,
        targetValue = 1f,
        animationSpec = infiniteRepeatable(
            animation = tween(1500, easing = LinearEasing),
            repeatMode = RepeatMode.Restart
        ),
        label = "scanLine"
    )

    LaunchedEffect(imageFile) {
        if (isProcessing) return@LaunchedEffect
        isProcessing = true
        
        ttsService.speakAsync("Analyse en cours")
        
        try {
            // 1. Face Detection
            val hasFace = FaceDetectionService.getInstance().hasFace(context, imageFile)
            
            if (hasFace) {
                delay(500)
                ttsService.speak("Attention. Ceci n'est pas un billet, c'est un visage. Veuillez scanner un billet.")
                delay(2000)
                navController.popBackStack() 
                return@LaunchedEffect
            }
            
            // 2. Currency Recognition
            val result = CurrencyRecognitionService.getInstance(context).recognizeCurrency(imageFile)
            
            delay(500)
            ttsService.stop()
            
            navController.currentBackStackEntry?.savedStateHandle?.set("result", result)
            navController.currentBackStackEntry?.savedStateHandle?.set("imagePath", imageFile.absolutePath)
            
            navController.navigate("results")
            
        } catch (e: Exception) {
            e.printStackTrace()
            ttsService.speak("Erreur lors de l'analyse. Veuillez reessayer.")
            delay(2000)
            navController.popBackStack()
        } finally {
            isProcessing = false
        }
    }

    Scaffold(
        containerColor = AppColors.BackgroundLight
    ) { paddingValues ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues)
                .padding(24.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Spacer(modifier = Modifier.weight(1f))

            // Image container
            Box(
                modifier = Modifier
                    .size(width = 320.dp, height = 420.dp)
                    .clip(RoundedCornerShape(24.dp))
                    .border(3.dp, AppColors.Accent, RoundedCornerShape(24.dp))
            ) {
                val bitmap = remember(imageFile) { BitmapFactory.decodeFile(imageFile.absolutePath) }
                if (bitmap != null) {
                    Image(
                        bitmap = bitmap.asImageBitmap(),
                        contentDescription = null,
                        modifier = Modifier.fillMaxSize(),
                        contentScale = ContentScale.Crop
                    )
                }

                Box(modifier = Modifier.fillMaxSize().background(Color.White.copy(alpha = 0.1f)))

                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(4.dp)
                        .align(Alignment.TopCenter)
                        .offset(y = (420.dp * scanLinePosition))
                        .background(
                             brush = Brush.horizontalGradient(
                                 colors = listOf(
                                     Color.Transparent,
                                     AppColors.Accent,
                                     AppColors.AccentLight,
                                     AppColors.Accent,
                                     Color.Transparent
                                 )
                             )
                        )
                )
            }
            
            Spacer(modifier = Modifier.height(50.dp))

            Card(
                colors = CardDefaults.cardColors(containerColor = AppColors.BackgroundCard),
                elevation = CardDefaults.cardElevation(defaultElevation = 5.dp),
                shape = RoundedCornerShape(20.dp),
                border = androidx.compose.foundation.BorderStroke(1.dp, AppColors.BorderColor)
            ) {
                Column(
                    modifier = Modifier.padding(28.dp),
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    Box(
                        modifier = Modifier
                            .size(55.dp)
                            .background(AppColors.Accent, CircleShape),
                        contentAlignment = Alignment.Center
                    ) {
                        Icon(Icons.Default.Search, contentDescription = null, tint = AppColors.TextLight)
                    }
                    
                    Spacer(modifier = Modifier.height(20.dp))
                    
                    Text(
                        text = "Analyse...",
                        style = MaterialTheme.typography.headlineMedium.copy(
                            fontWeight = FontWeight.Bold,
                            color = AppColors.TextDark
                        )
                    )
                    
                    Spacer(modifier = Modifier.height(16.dp))
                    
                    Text(
                        text = "Veuillez patienter",
                        style = MaterialTheme.typography.bodyLarge.copy(color = AppColors.TextGray)
                    )
                    
                    Spacer(modifier = Modifier.height(32.dp))
                    
                    LinearProgressIndicator(
                        modifier = Modifier.fillMaxWidth().height(5.dp).clip(RoundedCornerShape(4.dp)),
                        color = AppColors.Accent,
                        trackColor = AppColors.BorderColor
                    )
                }
            }

            Spacer(modifier = Modifier.weight(1f))

            TextButton(onClick = { navController.popBackStack() }) {
                Text(
                    text = "Annuler",
                    style = MaterialTheme.typography.bodyLarge.copy(color = AppColors.TextGray)
                )
            }
            
            Spacer(modifier = Modifier.height(20.dp))
        }
    }
}
